pkg/apis/core/validation/shoot.go

```diff
@@ -411,16 +410,26 @@ func validateAddons(addons *core.Addons, kubernetes core.Kubernetes, purpose *co
 }
 
 // ValidateNodeCIDRMaskWithMaxPod validates if the Pod Network has enough ip addresses (configured via the NodeCIDRMask on the kube controller manager) to support the highest max pod setting on the shoot
-func ValidateNodeCIDRMaskWithMaxPod(maxPod int32, nodeCIDRMaskSize int32) field.ErrorList {
+func ValidateNodeCIDRMaskWithMaxPod(maxPod int32, nodeCIDRMaskSize int32, networking core.Networking) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	free := float64(32 - nodeCIDRMaskSize)
+	bitLen := int32(32)
+	defaultNodeCIDRMaskSize := 24
+	if core.IsIPv6SingleStack(networking.IPFamilies) {
+		bitLen = 128
+		defaultNodeCIDRMaskSize = 64
+	}
+	bitLen -= nodeCIDRMaskSize
+
+	// Calculate how many addresses a single node podCIDR contains.
+	// This will overflow uint64 if nodeCIDRMaskSize <= 64 (default in IPv6), so use big.Int
```

Comment:
```suggestion
	// This will overflow uint64 if nodeCIDRMaskSize >= 64 (default in IPv6), so use big.Int
```

Reply:
I believe this logic is correct, I understand why this is confusing, though.
I added a few more comments and renamed variables to make this clearer (without changing the logic).